target/linux/ath79/dts/ar7100_mikrotik_routerboard-rb4xx.dtsi

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: GPL-1.0-or-later OR MIT
+
+// Based on ar71xx arch/mips/ath79/mach-rb4xx.c
+
+/dts-v1/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+#include "ar7100.dtsi"
+
+/ {
+	chosen {
+		bootargs = "console=ttyS0,115200";
+	};
+
+	aliases {
+		led-boot = &led_user_yellow;
+		led-failsafe = &led_user_yellow;
+		led-running = &led_user_yellow;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_user_yellow: user_yellow {
+			label = "rb4xx:yellow:user";
+			gpios = <&gpio 4 GPIO_ACTIVE_HIGH>;
+			default-state = "on";
+		};
+
+		led_1_green: 1_green {
+			label = "rb4xx:green:led1";
+			gpios = <&cpld 0 GPIO_ACTIVE_LOW>;
+		};
+
+		led_2_green: 2_green {
+			label = "rb4xx:green:led2";
+			gpios = <&cpld 2 GPIO_ACTIVE_LOW>;
+		};
+
+		led_3_green: 3_green {
+			label = "rb4xx:green:led3";
+			gpios = <&cpld 1 GPIO_ACTIVE_LOW>;
+		};
+
+		led_4_green: 4_green {
+			label = "rb4xx:green:led4";
+			gpios = <&cpld 3 GPIO_ACTIVE_LOW>;
+		};
+
+		led_5_green: 5_green {
+			label = "rb4xx:green:led5";
+			gpios = <&cpld 8 GPIO_ACTIVE_HIGH>;
+		};
+	};
+
+	keys: keys {
+		compatible = "gpio-keys";
+
+		status = "disabled";
+
+		reset_key: reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&gpio 7 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+	};
+
+	nand: nand {
+		compatible = "mikrotik,rb4xx-nand";
+	};
+};
+
+&uart {
+	status = "okay";
+};
+
+&gpio {
+	ngpios = <31>;
+};
+
+&pcie0 {
+	// NOT TESTED (n.b. ath79_pci_irq struct has the slot/pin/IRQ mappings if needed)
+	status = "okay";
+};
+
+&spi {
+	status = "okay";
+
+	compatible = "mikrotik,rb4xx-spi";
+
+	flash@0 {
+		compatible = "pm25lv512", "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <25000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "routerboot";
+				reg = <0x00000 0x0b000>;
+				read-only;
+			};
+
+			hard_config: partition@b000 {
+				label = "hard_config";
+				reg = <0x0b000 0x01000>;
+				read-only;
+			};
```

Comment:
What happened to 0xc000 to 0xd000?

Reply:
I pulled these values directly from https://github.com/openwrt/openwrt/blob/master/target/linux/ar71xx/files/arch/mips/ath79/mach-rb4xx.c#L143

Reply:
I don't question that, but still I'd like to know what's in that area.

Reply:
Mostly empty it seems. I mounted it as "mtd2" and dumped it out:

```
root@OpenWrt:/# hexdump -C /dev/mtd2
00000000  ff ff ff ff ff ff ff ff  ff ff 7d c7 00 00 00 00  |..........}.....|
00000010  5e 27 05 c6 0d 31 ea 5f  7a fd d9 f4 76 ea b2 3c  |^'...1._z...v..<|
00000020  cc 41 f4 d0 17 77 86 b5  ae 94 0d 5c 98 82 cb 85  |.A...w.....\....|
00000030  45 ed 16 d1 90 6b 1a 73  d1 6f 62 e4 81 34 0c eb  |E....k.s.ob..4..|
00000040  8c e5 5b 8d 23 1d 8e 34  8c 35 39 bf b3 c6 19 00  |..[.#..4.59.....|
00000050  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00000ff0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff 7f  |................|
00001000
```

Reply:
So, it's not empty.
Forum shows the same partitions but without providing info about the offsets:
https://openwrt.org/toh/mikrotik/rb493g#flash_layout

Can you reconfirm vendor partitioning yourself?

Reply:
I re-flashed the board with RouterOS vendor firmware, and it only shows me the NAND flash under partitions (and provides options to re-partition it). The NOR flash we are documenting here is not exposed.

The RooterBOOT bootloader also provides NAND visibility and repartition actions, but nothing for the NOR.

This NOR flash (as far as I can tell) holds the bootloader and it's config settings along with board specifics like serial number and mac addresses.

Reply:
This looks like what is usually referred to as "bios" on other RB devices. Not that we make any use of it anyway.

Reply:
So I have "bios" starting at 0x0d000, purely as that is where the old ar71xx build had it.

I believe the question is, is that the correct start point? (i.e. is there a better way to identify the correct partition sizes?)

Reply:
I'm currently working on a [parser for routerboot partitions](https://github.com/f00b4r0/openwrt/commits/rbfixes).
There's no proper partition scheme, it's a mix of fixed and dynamic ones that differ between routerboard platforms. Some partitions can be identified (hard_config, soft_config), others cannot. "bios" typically cannot, as far as I know. Then again, the only partitions effectively used by OpenWRT are "hard_config" and "soft_config". The other ones are mostly a cosmetic matter.

Reply:
That is a very interesting solution!

> Could you provide a full dump of the NOR chip somewhere so I can take a look?

Sure. Do you want me to just 'dd' it into a file and upload somewhere?

Reply:
A concatenation won't do since your DTS has holes; I would recommend adding a top level "RouterBoot" partition container as is done [here](https://github.com/f00b4r0/openwrt/blob/e4ce3109f28712dad3b3db27938a7b853346d9b6/target/linux/ramips/dts/mt7621_mikrotik_rbm11g.dts#L94) (in your case it would cover the entire flash area presumably) and dump that top level block with dd.
Thanks!

Reply:
Ok, I've dumped it out (and placed that on another repo you have an invite for).

Reply:
Hi, invite not working. Can you mail it to me instead? Thanks!

Reply:
OK so after taking a closer look, it seems you have a bootloader at 0x0000 size 0xb000, hard_config at 0xb000 size 0x1000, "bios" at 0xc000 size 0x1000, what looks like some bootloader code at 0xd000 size 0x1000, and soft_config at 0xf000 size 0x1000

HTH

Reply:
Thanks. Looks like moving the "bios" section back to 0xc000 and shrinking it, adding a new "bootloader" partition at 0xd000, and then leave the space from 0xe000 to oxf000 undefined, is correct?

Reply:
Sorry, taking into account FF padding, the second bootloader part extends up to 0xf000, so size 0x2000.

Reply:
Nice, now the entire address space will then be defined and mounted!

BTW, is there a common name in use for this 2nd bootloader partition?

Reply:
Usually "routerboot2", assuming it is indeed a bootloader :)

Reply:
and when I say "extends" I really mean "can extend" since FF fill means the flash is blank there :)

Reply:
This device does have a backup "protected" bootloader... I'm unclear where that lives!

Reply:
Either it's that or on the nand. I'm not very familiar with NAND-based routerboards :)

Reply:
I'm going to "resolve" this thread as the entire NOR is now mapped out.

I'm also looking forward to using the new auto-parser you have, and then I can rip this whole section out once merged!